src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsInAnyOrderIT.java

```diff
@@ -0,0 +1,735 @@
+import static io.sphere.sdk.models.LocalizedString.ofEnglish;
+import static io.sphere.sdk.utils.SphereInternalUtils.asSet;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singleton;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ProductSyncWithReferencedProductsInAnyOrderIT {
+    private static ProductType productType;
+
+
+    private ProductSyncOptions syncOptions;
+    private Product product;
+    private List<String> errorCallBackMessages;
+    private List<String> warningCallBackMessages;
+    private List<Throwable> errorCallBackExceptions;
+    private List<UpdateAction<Product>> actions;
+
+    @BeforeAll
+    static void setup() {
+        deleteProductSyncTestData(CTP_TARGET_CLIENT);
+        productType = createProductType(PRODUCT_TYPE_RESOURCE_PATH, CTP_TARGET_CLIENT);
+    }
+
+    @BeforeEach
+    void setupTest() {
+        clearSyncTestCollections();
+        deleteAllProducts(CTP_TARGET_CLIENT);
+        syncOptions = buildSyncOptions();
+
+        final ProductDraft productDraft = ProductDraftBuilder
+            .of(productType, ofEnglish("foo"), ofEnglish("foo-slug"), emptyList())
+            .masterVariant(ProductVariantDraftBuilder.of().key("foo").sku("foo").build())
+            .key("foo")
+            .build();
+
+        product = executeBlocking(CTP_TARGET_CLIENT.execute(ProductCreateCommand.of(productDraft)));
+    }
+
+    private void clearSyncTestCollections() {
+        errorCallBackMessages = new ArrayList<>();
+        errorCallBackExceptions = new ArrayList<>();
+        warningCallBackMessages = new ArrayList<>();
+        actions = new ArrayList<>();
+    }
+
+    private ProductSyncOptions buildSyncOptions() {
+        final Consumer<String> warningCallBack = warningMessage -> warningCallBackMessages.add(warningMessage);
+
+        return ProductSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+                                        .errorCallback(this::collectErrors)
+                                        .beforeUpdateCallback(this::collectActions)
+                                        .warningCallback(warningCallBack)
+                                        .build();
+    }
+
+    private void collectErrors(@Nullable final String errorMessage, @Nullable final Throwable exception) {
+        errorCallBackMessages.add(errorMessage);
+        errorCallBackExceptions.add(exception);
+    }
+
+    private List<UpdateAction<Product>> collectActions(@Nonnull final List<UpdateAction<Product>> actions,
+                                @Nonnull final ProductDraft productDraft,
+                                @Nonnull final Product product) {
+        this.actions.addAll(actions);
+        return actions;
+    }
+
+    @AfterAll
+    static void tearDown() {
+        deleteProductSyncTestData(CTP_TARGET_CLIENT);
+    }
+
+    @Test
+    void sync_withSingleChildDraftSuppliedBeforeParent_shouldSyncCorrectly() {
```

Comment:
Do you verify that custom object is being cleaned up correctly?

Reply:
f2dc950